Upgrade tooling for 9.0x

radical opened the pull request:
- Update dependencies from `.NET 9 Eng` channel
- Update nuget.config to use `dotnet9` nuget feed instead of `nuget8`
- Update tests to run with 8.0 runtime
- Resolve new xunit analyzer warnings

All the projects still target `net8.0`.

Future TODO:
- add 9.0 templates, and tests

Fixes https://github.com/dotnet/aspire/issues/5446 .

TODO before merge:
- [x] check that code coverage report is unaffected
- [x] Number of tests is the same as a `main` run
- [ ] Diff package build output to ensure contents of packages are unaffected (added by @joperezr)
- [x] Validate with an internal build

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/5483)

joperezr commented on src/Aspire.Dashboard/Otlp/Storage/ApplicationKey.cs:
```diff
@@ -16,7 +16,7 @@ public int CompareTo(ApplicationKey other)
     }
 
-    public bool EqualsCompositeName(string name)
+    public bool EqualsCompositeName(string? name)
```
This is technically a breaking change, although because it is in the Dashboard we don't really care. For other src projects with Nullability issues (if any) I think it'd be best to just baseline the warnings and/or turn off nullability than trying to fix it on this PR.

joperezr commented on tests/Aspire.Hosting.Testing.Tests/TestingBuilderTests.cs:
```diff
@@ -230,7 +230,7 @@ public async Task CrashTests(bool genericEntryPoint, string crashArg)
             genericEntryPoint ? DistributedApplicationTestingBuilder.CreateAsync<Projects.TestingAppHost1_AppHost>([$"--crash-{crashArg}"], cts.Token).WaitAsync(cts.Token)
                                 : DistributedApplicationTestingBuilder.CreateAsync(typeof(Projects.TestingAppHost1_AppHost), [$"--crash-{crashArg}"], cts.Token).WaitAsync(cts.Token))
-                .ConfigureAwait(false);
+;
```
NIT: Fix formatting by removing new-line.

radical replied:
I'll do a cleanup pass once all the tests are passing.

joperezr commented on tests/Aspire.Workload.Tests/StarterTemplateRunTestsBase.cs:
```diff
@@ -103,7 +103,7 @@ static async Task<List<string[]>> GetAndValidateCellTexts(ILocator tableLoc)
                     r => Assert.True(int.TryParse(r, out var actualTempC) && actualTempC >= -20 && actualTempC <= 55),
                     r => Assert.True(int.TryParse(r, out var actualTempF) && actualTempF >= -5 && actualTempF <= 133),
-                    r => Assert.Contains(r, ["Freezing", "Bracing", "Chilly", "Cool", "Mild", "Warm", "Balmy", "Hot", "Sweltering", "Scorching"]));
+                    r => Assert.Contains(r, new HashSet<string>(){"Freezing", "Bracing", "Chilly", "Cool", "Mild", "Warm", "Balmy", "Hot", "Sweltering", "Scorching"}));
```
Interesting, I would have expected this to work with the new collection literals.

radical replied:
IIRC it was not being able to resolve the overload between `HashSet<>` and `SortedSet<>`.

joperezr commented on tests/Shared/WorkloadTesting/data/nuget8.config:
```diff
@@ -9,7 +9,7 @@
     <add key="dotnet-public" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet-public/nuget/v3/index.json" />
     <add key="dotnet-eng" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet-eng/nuget/v3/index.json" />
-    <add key="dotnet8" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet8/nuget/v3/index.json" />
+    <add key="dotnet9" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet9/nuget/v3/index.json" />
```
So we don't need dotnet8 feed anymore? I think we still push our packages there so we have to update our default channels in darc.

radical replied:
But we shouldn't be fetching any of our packages from these feeds (except `BUILT_NUGETS_PATH`) for tests. Or am I missing something?

joperezr commented on tests/Shared/Aspire.Workload.Testing.targets:
```diff
@@ -19,6 +19,9 @@
   <ItemGroup>
     <PackageReference Include="Microsoft.NET.Runtime.WorkloadTesting.Internal" />
+
+    <AdditionalSharedFrameworkToInstallArguments Include="-Runtime dotnet -Version 8.0.8 -Channel 8.0" />
```
Is there a way to reuse the dotnet installed by the original build.cmd/sh call so we don't have to add this? I want to avoid needing to update a number in several places (runtime framework in this case 8.0.8) as those change each month.

radical replied:
This is specifically for workload testing. It will be much cleaner to use the existing support for this. But we could try to use the restored `.dotnet` also, but also need to account for `.dotnet` not being installed when the system `dotnet` is sufficient.

joperezr commented on tests/helix/send-to-helix-inner.proj:
```diff
@@ -8,13 +8,15 @@
 
     <IncludeDotNetCli>true</IncludeDotNetCli>
     <DotNetCliPackageType>sdk</DotNetCliPackageType>
+    <DotNetCliVersion>8.0.401</DotNetCliVersion>
```
How will this work for template tests for 9.0 templates? won't they require a 9.0 SDK?

radical replied:
Still going through a first pass to get tests passing :D

radical replied:
This will require us to bundle the sdk ourselves, because arcade doesn't seem to support additional runtimes.
To confirm that this works I think this PR should include a couple of tests where the test project targets net9.0 (best in workload tests).

joperezr commented on Directory.Build.props:
```diff
@@ -3,6 +3,7 @@
   <Import Project="$(RepositoryEngineeringDir)/apicompat/PublicApiAnalyzer.props" />
 
   <PropertyGroup>
+    <NetCurrent>net8.0</NetCurrent>
```
Is this property used anywhere else in arcade or is it solely used by consuming repos? my concern is that changing this in order to avoid having to modify individual projects might unintentionally affect how the infrastructure work.

radical replied:
Good point. I can drop this in the cleanup.

eerhardt replied:
Looks like efcore calls this `DefaultNetCoreTargetFramework`.

https://github.com/dotnet/efcore/blob/ecfee78eb1fa2b2eaa0dbf945f1d4f8fa571be74/src/EFCore.Abstractions/EFCore.Abstractions.csproj#L5C24-L5C53

eerhardt commented on global.json:
```diff
@@ -1,16 +1,30 @@
+        "8.0.8"
+      ],
+      "dotnet/arm64": [
+        "8.0.8"
```
Do we have to have this duped by architecture? Can we do what efcore does?

https://github.com/dotnet/efcore/blob/ecfee78eb1fa2b2eaa0dbf945f1d4f8fa571be74/global.json#L9-L12

radical committed:
Update dependencies from ".NET 9 Eng" channel

Updating 'Microsoft.SourceBuild.Intermediate.source-build-reference-packages': '8.0.0-alpha.1.23516.4' => '9.0.0-alpha.1.24453.2' (from build '20240903.2' of 'https://github.com/dotnet/source-build-reference-packages')
Updating 'Microsoft.SourceBuild.Intermediate.arcade': '8.0.0-beta.24426.2' => '9.0.0-beta.24453.1' (from build '20240903.1' of 'https://github.com/dotnet/arcade')
Updating 'Microsoft.DotNet.Arcade.Sdk': '8.0.0-beta.24426.2' => '9.0.0-beta.24453.1' (from build '20240903.1' of 'https://github.com/dotnet/arcade')
Updating 'Microsoft.DotNet.Build.Tasks.Installers': '8.0.0-beta.24426.2' => '9.0.0-beta.24453.1' (from build '20240903.1' of 'https://github.com/dotnet/arcade')
Updating 'Microsoft.DotNet.Build.Tasks.Workloads': '8.0.0-beta.24426.2' => '9.0.0-beta.24453.1' (from build '20240903.1' of 'https://github.com/dotnet/arcade')
Updating 'Microsoft.DotNet.Helix.Sdk': '8.0.0-beta.24426.2' => '9.0.0-beta.24453.1' (from build '20240903.1' of 'https://github.com/dotnet/arcade')
Updating 'Microsoft.DotNet.RemoteExecutor': '8.0.0-beta.24426.2' => '9.0.0-beta.24453.1' (from build '20240903.1' of 'https://github.com/dotnet/arcade')
Updating 'Microsoft.DotNet.SharedFramework.Sdk': '8.0.0-beta.24426.2' => '9.0.0-beta.24453.1' (from build '20240903.1' of 'https://github.com/dotnet/arcade')
Updating 'Microsoft.DotNet.XUnitExtensions': '8.0.0-beta.24426.2' => '9.0.0-beta.24453.1' (from build '20240903.1' of 'https://github.com/dotnet/arcade')

radical committed:
Add dotnet 8 versions for testing

radical committed:
[tests] Install 8.0 runtime for workload testing

radical committed:
Resolve xUnit2013 warning

`Do not use Assert.Equal() to check for collection size. Use
Assert.Single instead.xUnit2013`

radical committed:
Resolve xUnit2029 warning

`Do not use Assert.Empty() to check if a value does not exist in a
collection. Use Assert.DoesNotContain() instead.xUnit2029`

radical committed:
Track API change in EqualException

radical committed:
Ignore warning xUnit1030

`error xUnit1030: Test methods should not call ConfigureAwait(false), as
it may bypass parallelization limits. Omit ConfigureAwait, or use
ConfigureAwait(true) to avoid CA2007.
(https://xunit.net/xunit.analyzers/rules/xUnit1030)`

radical committed:
Fix xUnit1012 analyzer warning.

Warning of the form:
`error xUnit1012: Null should not be used for type parameter
'secondaryApiKey' of type 'string'. Use a non-null value, or convert the
parameter to a nullable type.
(https://xunit.net/xunit.analyzers/rules/xUnit1012)`

radical committed:
Fix compile error.

`error CS0121: The call is ambiguous between the following methods or
properties: 'Assert.Contains<T>(T, HashSet<T>)' and
'Assert.Contains<T>(T, SortedSet<T>)'`

radical committed:
couple more fixes

radical committed:
Update nuget.config to use dotnet9 sources instead of dotnet8

radical commented:
/azp run

radical committed:
Remove unnecessary package reference to Microsoft.DotNet.XunitAssert

radical committed:
Address xUnit1030 warning.

`error xUnit1030: Test methods should not call ConfigureAwait(false), as it may bypass parallelization limits. Omit ConfigureAwait, or use ConfigureAwait(true) to avoid CA2007. (https://xunit.net/xunit.analyzers/rules/xUnit1030)`

radical committed:
Ignore some lint warnings for eng/common/template-guidance.md from arcade

radical committed:
cleanup

radical commented:
TODO:
- [ ] check that code coverage report is unaffected
- [ ] Number of tests is the same as a `main` run

radical commented on eng/Version.Details.xml:
```diff
@@ -114,56 +114,50 @@
+      <Sha>dd332f2d4e21daa8b79f84251ab156af9a0b11b2</Sha>
       <SourceBuild RepoName="arcade" ManagedOnly="true" />
     </Dependency>
-    <!-- Intermediate is necessary for source build. -->
-    <Dependency Name="Microsoft.SourceBuild.Intermediate.xliff-tasks" Version="1.0.0-beta.23475.1" CoherentParentDependency="Microsoft.DotNet.Arcade.Sdk">
-      <Uri>https://github.com/dotnet/xliff-tasks</Uri>
-      <Sha>73f0850939d96131c28cf6ea6ee5aacb4da0083a</Sha>
-      <SourceBuild RepoName="xliff-tasks" ManagedOnly="true" />
-    </Dependency>
```
From the first commit - does dropping this seem correct? The assembly has moved to `dotnet/arcade`.

joperezr replied:
I think that is fine, and I think you can also drop the dependency over from lines 152-155 above for that matter (not sure why we had both), but shouldn't we have now a new dependency coming from arcade for it?

radical replied:
We'll need to add a dependency like `dotnet/runtime` has:
```xml
    <Dependency Name="Microsoft.DotNet.XliffTasks" Version="9.0.0-beta.24423.2">
      <Uri>https://github.com/dotnet/arcade</Uri>
      <Sha>91599268652b51969b8d8088d4f2f2ba7b3ebb19</Sha>
    </Dependency>
```

radical replied:
added that here.

radical commented:
Read commit by commit to review this. The first commit is the biggest and is the update of the dependencies, mostly from arcade.

radical commented:
/azp run

radical committed:
Merge remote-tracking branch 'origin/main' into switch-9.0

# Conflicts:
#	tests/Aspire.Elastic.Clients.Elasticsearch.Tests/Aspire.Elastic.Clients.Elasticsearch.Tests.csproj

radical committed:
Fix AzureFunctionsEndToEnd

radical committed:
Track changes for Aspire.Hosting.Azure.Functions

radical committed:
Fix CA2007: src\Aspire.Hosting\Health\ResourceHealthCheckScheduler.cs(38,45): error CA2007: Consider calling ConfigureAwait on the awaited task

joperezr commented on eng/Versions.props:
```diff
@@ -12,6 +12,12 @@
     <BaselineVersionForPackageValidation>8.0.1</BaselineVersionForPackageValidation>
+    <DefaultTargetFramework>net8.0</DefaultTargetFramework>
+    <!-- dotnet 8.0 versions for running tests -->
+    <DotNetRuntimeVersionForTesting>8.0.8</DotNetRuntimeVersionForTesting>
```
I'm wondering how we can keep this one up-to-date. One option would be if we use the `Microsoft.NetCore.App.Ref` package and dependency flow for the version as we know that one will push the update to us. Anyway, no need to fix it now but we should consider how to update this (and the sdk below)

radical replied:
`dotnet/runtime` is using `Microsoft.DotNet.ApiCompat.Task` for that. Either should work well.

joperezr commented on tests/Aspire.Workload.Tests/StarterTemplateRunTestsBase.cs:
```diff
@@ -103,7 +103,7 @@ static async Task<List<string[]>> GetAndValidateCellTexts(ILocator tableLoc)
                     r => Assert.True(int.TryParse(r, out var actualTempC) && actualTempC >= -20 && actualTempC <= 55),
                     r => Assert.True(int.TryParse(r, out var actualTempF) && actualTempF >= -5 && actualTempF <= 133),
-                    r => Assert.Contains(r, ["Freezing", "Bracing", "Chilly", "Cool", "Mild", "Warm", "Balmy", "Hot", "Sweltering", "Scorching"]));
+                    r => Assert.Contains(r, new HashSet<string>{"Freezing", "Bracing", "Chilly", "Cool", "Mild", "Warm", "Balmy", "Hot", "Sweltering", "Scorching"}));
```
Still curious why this didn't work with the collection initializer, but oh well, no need to change it.

radical replied:
`error CS0121: The call is ambiguous between the following methods or
properties: 'Assert.Contains<T>(T, HashSet<T>)' and
'Assert.Contains<T>(T, SortedSet<T>)'`

joperezr commented on tests/helix/send-to-helix-inner.proj:
```diff
@@ -13,8 +13,7 @@
 
-    <GlobalJsonContent>$([System.IO.File]::ReadAllText('$(RepoRoot)global.json'))</GlobalJsonContent>
-    <DotNetCliVersion>$([System.Text.RegularExpressions.Regex]::Match($(GlobalJsonContent), '(%3F&lt;="dotnet": ").*(%3F=")'))</DotNetCliVersion>
+    <DotNetCliVersion>$(DotNetSdkVersionForTesting)</DotNetCliVersion>
```
We'll want to make sure now that we update that property so that it at least equals the one in global.json. I wonder if we can add a target somewhere in our build that detects if we've fallen behind and if so to fail the build...

No need to add that in this change though, don't block this PR for that.

radical replied:
Once we get the version from a `dotnet/sdk` dependency, it should always(?) be newer than `global.json`.

joperezr commented on global.json:
```diff
@@ -1,16 +1,30 @@
 {
   "sdk": {
-    "version": "8.0.303",
+    "version": "9.0.100-preview.7.24407.12",
```
This is fine for now. We'll want to bump to rc1 that got shipped today soon.

joperezr reviewed (approved):
Changes look great @radical thanks for working on this! Let's prioritize getting this merged given the breath as otherwise it is likely to get conflicts with other PRs.

radical committed:
Merge remote-tracking branch 'origin/main' into switch-9.0

radical committed:
Add dependency to Microsoft.DotNet.XliffTasks from dotnet/arcade

radical commented:
Issue to track follow up work - https://github.com/dotnet/aspire/issues/5676

radical commented:
Internal build for validation - https://dev.azure.com/dnceng/internal/_build/results?buildId=2535499

radical committed:
Add back skipping hosting.sdk, and projectTemplates packages

.. in workload testing as it is required for the internal build.

`Aspire.Hosting.Sdk.Msi.arm64.9.0.0-preview.4.24460.2.nupkg, Aspire.Hosting.Sdk.Msi.x64.9.0.0-preview.4.24460.2.nupkg, Aspire.Hosting.Sdk.Msi.x86.9.0.0-preview.4.24460.2.nupkg, Aspire.ProjectTemplates.Msi.arm64.9.0.0-preview.4.24460.2.nupkg, Aspire.ProjectTemplates.Msi.x64.9.0.0-preview.4.24460.2.nupkg, Aspire.ProjectTemplates.Msi.x86.9.0.0-preview.4.24460.2.nupkg`

radical committed:
[tests] Increase the default timeout from 10mins to 15mins for basictests

.. like the various hosting integration tests. The run time can vary
especially if docker needs to fetch an image.
